### PR TITLE
core: Implement TypedAttribute, leverage in declarative assembly format.

### PR DESCRIPTION
**Pull request by PapyChacal**
Implement TypedAttribute, a baseclass for attributes having a type attribute :upside_down_face: Mirroring the MLIR interface.
This is useful for the declarative assembly format, where when referencing an integer of a known type, we want to parse and print it with its type elided.

### CODE REVIEW

File: xdsl/irdl/declarative_assembly_format.py [hunk 1]
```diff
@@ -727,18 +728,31 @@ class AttributeVariable(FormatDirective):
     """Should this attribute be put in the attribute or property dictionary."""
     unique_base: type[Attribute] | None
     """The known base class of the Attribute, if any."""
+    unique_type: Attribute | None
```
Comment (math-fehr) [inline xdsl/irdl/declarative_assembly_format.py:731]: Can you add the documentation for the field?

File: xdsl/irdl/declarative_assembly_format_parser.py [hunk 2]
```diff
@@ -371,15 +373,32 @@ def parse_optional_variable(
...
                 # This is technically an approximation, but appears to be good enough
                 # for xDSL right now.
-                if unique_base is not None and unique_base in Builtin.attributes:
+                unique_type = None
+                if unique_base is not None and issubclass(unique_base, TypedAttribute):
+                    constr = attr_def.constr
+                    if isinstance(constr, ParamAttrConstraint):
+                        type_constraint = constr.param_constrs[unique_base.type_index]
+                        if type_constraint.can_infer(set()):
+                            unique_type = type_constraint.infer({})
```
Comment (math-fehr) [inline xdsl/irdl/declarative_assembly_format_parser.py:382]: Note, we should improve this to allow things like `AllOf(ParamAttrConstraint(...), ...)` as well?

Comment (math-fehr) [reply xdsl/irdl/declarative_assembly_format_parser.py:382]: By improve, I mean adding a comment and an issue for now :)
If you can, also adding the issue number in the comment as well

File: xdsl/irdl/irdl.py [hunk 3]
```diff
@@ -2398,6 +2402,18 @@ def field(self: _PAttrT):
     for idx, (param_name, _) in enumerate(attr_def.parameters):
         new_fields[param_name] = param_name_field(idx)
 
+    if issubclass(cls, TypedAttribute):
```
Comment (math-fehr) [inline xdsl/irdl/irdl.py:2405]: Very nice

File: xdsl/irdl/irdl.py [hunk 4]
```diff
@@ -2285,6 +2286,9 @@ def irdl_param_attr_get_param_type_hints(cls: type[_PAttrT]) -> list[tuple[str,
     for field_name, field_type in get_type_hints(cls, include_extras=True).items():
         if field_name == "name" or field_name == "parameters":
             continue
+        # Allow type_index as defined by TypedAttribute
```
Comment (math-fehr) [inline xdsl/irdl/irdl.py:2289]: I'm wondering if `type_index` should not be a method or property instead, just to make this part less weird?

**Review by math-fehr — COMMENTED** [review]
Really nice! I added some comments

File: xdsl/irdl/declarative_assembly_format_parser.py (hunk 2 above)
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format_parser.py:382]: I don't get it
Do you mean improve the .infer to have this working in more cases?

File: xdsl/ir/core.py [hunk 5]
```diff
@@ -666,6 +666,25 @@ def _verify(self):
         super()._verify()
 
 
+class TypedAttribute(ParametrizedAttribute, Generic[AttributeInvT], ABC):
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/ir/core.py:669]: I like the field option but no strong feeling; want me to make it a method?

File: xdsl/irdl/irdl.py [hunk 6]
```diff
@@ -1064,7 +1065,7 @@ def opt_result_def(
 
 
 def prop_def(
-    constraint: type[AttributeInvT] | TypeVar,
+    constraint: type[AttributeInvT] | TypeVar | AttrConstraint,
```
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/irdl/irdl.py:1068]: :shrug: 
He didn't tell me
I think a lot of that part of the codebase needs some refreshing on that side either way, can we just be happy he's happy for now?

File: xdsl/ir/core.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (PapyChacal) [reply xdsl/ir/core.py:669]: Switched to method

File: xdsl/irdl/irdl.py (hunk 4 above)
Comment (PapyChacal) [reply xdsl/irdl/irdl.py:2289]: Done

File: xdsl/irdl/declarative_assembly_format.py [hunk 7]
```diff
@@ -727,18 +728,32 @@ class AttributeVariable(FormatDirective):
...
+            attr = unique_base.new(unique_base.parse_parameter(parser))
         else:
-            if issubclass(self.unique_base, ParametrizedAttribute):
-                attr = self.unique_base.new(self.unique_base.parse_parameters(parser))
-            elif issubclass(self.unique_base, Data):
-                unique_base = cast(type[Data[Any]], self.unique_base)
-                attr = unique_base.new(unique_base.parse_parameter(parser))
-            else:
-                raise ValueError("Attributes must be Data or ParameterizedAttribute!")
+            raise ValueError("Attributes must be Data or ParameterizedAttribute!")
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format.py:756]:
```suggestion
            raise ValueError("Attributes must be Data or ParameterizedAttribute")
```

File: xdsl/irdl/declarative_assembly_format_parser.py [hunk 8]
```diff
@@ -371,15 +373,35 @@ def parse_optional_variable(
                 # Always qualify builtin attributes
                 # This is technically an approximation, but appears to be good enough
                 # for xDSL right now.
-                if unique_base is not None and unique_base in Builtin.attributes:
+                unique_type = None
+                if unique_base is not None and issubclass(unique_base, TypedAttribute):
+                    constr = attr_def.constr
+                    # TODO: generalize.
```
Comment (superlopuh) [inline xdsl/irdl/declarative_assembly_format_parser.py:379]: Can you please expand on this comment, and file an issue?

File: xdsl/irdl/irdl.py [hunk 9]
```diff
@@ -2354,6 +2356,28 @@ def from_pyrdl(pyrdl_def: type[ParametrizedAttribute]) -> ParamAttrDef:
...
+            except ValueError:
+                raise PyRDLAttrDefinitionError(
+                    f"TypedAttribute {pyrdl_def.__name__} should have a 'type' parameter."
+                )
+            typed_hint = param_hints[type_index][1]
+            if get_origin(typed_hint) is Annotated:
+                typed_hint = get_args(typed_hint)[0]
+            # import pdb
+
+            # pdb.set_trace()
```
Comment (superlopuh) [inline xdsl/irdl/irdl.py:2373]:
```suggestion
            
```

**Review by superlopuh — APPROVED** [review]

File: xdsl/irdl/declarative_assembly_format.py (hunk 7 above)
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format.py:756]: :open_mouth:

File: xdsl/irdl/declarative_assembly_format_parser.py (hunk 8 above)
Comment (PapyChacal) [reply xdsl/irdl/declarative_assembly_format_parser.py:379]: https://github.com/xdslproject/xdsl/Done: issues/2499

**Pull request merged**